Maple2.Server.Game/Model/Field/Actor/ActorState/AiState.cs

```diff
@@ -0,0 +1,452 @@
+﻿using Maple2.Model.Enum;
+using Maple2.Model.Metadata;
+using Maple2.Server.Core.Constants;
+using Maple2.Server.Game.Model.Skill;
+using Maple2.Server.Game.Packets;
+using Microsoft.EntityFrameworkCore.Metadata.Conventions;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Numerics;
+using System.Text;
+using System.Threading.Tasks;
+using static Maple2.Model.Metadata.AiMetadata;
+using static System.Runtime.InteropServices.JavaScript.JSType;
+
+namespace Maple2.Server.Game.Model.Field.Actor.ActorState;
+
+public class AiState {
+    private readonly FieldNpc actor;
+    public AiMetadata? AiMetadata { get; private set; }
+    private Node? battle = null;
+
+    private List<StackEntry> aiStack = new List<StackEntry>();
+
+    public AiState(FieldNpc actor) {
+        this.actor = actor;
+    }
+
+    [MemberNotNullWhen(true, "AiMetadata")]
+    public bool SetAi(string name) {
+        if (name == string.Empty) {
+            AiMetadata = null;
+
+            return false;
+        }
+
+        AiMetadata? metadata;
+
+        if (!actor.Field.AiMetadata.TryGet(name, out metadata)) {
+            return false;
+        }
+
+        AiMetadata = metadata;
+
+        battle = new Node("battle", AiMetadata.Battle);
+
+        return true;
+    }
+
+    private long nextUpdate = 0;
+    private AiMetadata? lastEvaluated = null;
+
+    public void Update(long tickCount) {
+        if (nextUpdate == 0) {
+            nextUpdate = tickCount + 1000;
+        }
+
+        if (nextUpdate > tickCount) {
+            return;
+        }
+        if (AiMetadata is null) {
+            if (actor.Value.Metadata.AiPath != "") {
+                actor.AppendDebugMessage("Missing AI\n");
+                actor.AppendDebugMessage(actor.Value.Metadata.AiPath);
+            }
+
+            aiStack.Clear();
+
+            return;
+        }
+
+
+        if (aiStack.Count == 0) {
+            if (AiMetadata.Battle.Length == 0) {
+                return;
+            }
+
+            if (battle is not null) {
+                Push(battle);
+            }
+        }
+
+        while (aiStack.Count > 0) {
+            StackEntry entry = aiStack.Last();
+
+            int index = entry.Index;
+            int last = aiStack.Count - 1;
+
+            if (index >= entry.Node.Entries.Length) {
+                aiStack.RemoveAt(last);
+
+                continue;
+            }
+
+            int nextIndex = entry.LockIndex ? entry.Node.Entries.Length : index + 1;
+
+            aiStack[last] = new StackEntry() { Node = entry.Node, Index = nextIndex, LockIndex = entry.LockIndex };
+
+            Process(entry.Node.Entries[index]);
+
+            break;
+        }
+
+        nextUpdate = tickCount + 1000;
```

Comment:
Can you comment as to why you're using a 1000 tick add?

Reply:
Added a comment
`arbitrary delay to process 1 actionable node at a time to get something up and running.`
`almost certainly wrong, and probably depends on node properties & arbitrary task durations, and probably a value in the server constants.`